### PR TITLE
[Divulgence pruning] Conformance tests implementation [DPP-484]

### PR DESCRIPTION
**Pull request by tudor-da**
Implements conformance tests for asserting the state of divulgence pruning in the `participant-integration-api`:
* The current test implementation asserts that divulgence events are not pruned
* A commented assertion, marked with **TODO** in the tests, asserts the future expectation when performing divulgence pruning: all divulgence and disclosure events are pruned up to the pruned offset.

CHANGELOG_BEGIN
* integration-kit: extended the Ledger API test tool with tests for the pruning of all divulgence events.

CHANGELOG_END

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala [hunk 1]
```diff
@@ -526,6 +529,127 @@ class ParticipantPruningIT extends LedgerTestSuite {
...
+    "PRRetroactiveDivulgences",
+    "Divulgence pruning succeeds",
+    allocate(SingleParty, SingleParty),
+    runConcurrently = false,
+  )(implicit ec => { case Participants(Participant(alpha, alice), Participant(beta, bob)) =>
+    for {
+      contract <- alpha.create(alice, Contract(alice))
+      divulgenceHelper <- alpha.create(alice, DivulgenceHelper(alice, bob))
+
+      divulgence <- beta.exerciseAndGetContract[Divulgence](bob, divulgenceHelper.exerciseSign)
```
Comment (meiersi-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala:542]: Share this code across test-cases.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala [hunk 2]
```diff
@@ -526,6 +529,127 @@ class ParticipantPruningIT extends LedgerTestSuite {
     }
   })
 
+  test(
+    "PRRetroactiveDivulgences",
+    "Divulgence pruning succeeds",
+    allocate(SingleParty, SingleParty),
+    runConcurrently = false,
```
Comment (meiersi-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala:536]:
```suggestion
    runConcurrently = false,  // pruning call may interact with other tests
```

File: ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  with
+    divulger: Party
+    divulgee: Party
+  where
+    signatory divulger
+    controller divulger can
+      nonconsuming DivulgeNoDisclose: ()
+        with divulgenceHelper: ContractId Divulgence
+          do contract <- create Contract with party=divulger
+             exercise divulgenceHelper Divulge with contractId=contract
```
Comment (meiersi-da) [inline ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml:54]:
```suggestion
        with divulgence: ContractId Divulgence
          do contract <- create Contract with party=divulger
             exercise divulgence Divulge with contractId=contract
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala [hunk 4]
```diff
@@ -526,6 +529,127 @@ class ParticipantPruningIT extends LedgerTestSuite {
...
+  )(implicit ec => { case Participants(Participant(alpha, alice), Participant(beta, bob)) =>
+    for {
+      divulgenceHelper <- alpha.create(alice, DivulgenceHelper(alice, bob))
+
+      divulgence <- beta.exerciseAndGetContract[Divulgence](bob, divulgenceHelper.exerciseSign)
+
+      divulgeNotDiscloseTemplate <- alpha.create(alice, DivulgeNotDiscloseTemplate(alice, bob))
+
+      // Alice creates and divulges without disclosure a contract to Bob
```
Comment (meiersi-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala:567]:
```suggestion
      // Alice creates contract in a context not visible to Bob and follows with a divulgence to Bob in the same transaction 
```

File: ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+             return ()
+
+template Dummy
```
Comment (meiersi-da) [inline ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml:57]:
```suggestion
// Dummy contracts are used to signal transactions on the TransactionTree service.
// It is structurally equal to the `Contract` template, but kept separate for the signalling purpose.
template Dummy
```

File: ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  where
+    signatory divulgee, divulger
+    controller divulger can
+      nonconsuming Divulge: ()
+        with contractId: ContractId Contract
+          -- create an event so the transaction is visible on the GetTransactionTrees endpoint
+          do create Dummy with holder=divulger
+             fetch contractId
+             return ()
+      nonconsuming CreateAndDisclose: ContractId Contract
```
Comment (meiersi-da) [inline ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml:25]:
```suggestion
      
      nonconsuming CreateAndDisclose: ContractId Contract
```

File: ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+          do create Contract with party=divulger
+    controller divulgee can
```
Comment (meiersi-da) [inline ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml:27]:
```suggestion

    controller divulgee can
```

File: ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+  where
+    signatory party
+
+template Divulgence
+  with
+    divulgee: Party
+    divulger: Party
+  where
+    signatory divulgee, divulger
+    controller divulger can
```
Comment (meiersi-da) [inline ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml:18]:
```suggestion

    controller divulger can
```

File: ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+      nonconsuming CreateAndDisclose: ContractId Contract
+          do create Contract with party=divulger
+    controller divulgee can
+      nonconsuming CanFetch: ContractId Dummy
+        with contractId: ContractId Contract
+          do fetch contractId
+             -- create an event so the transaction is visible on the GetTransactionTrees endpoint
+             create Dummy with holder=divulgee
+
+template DivulgenceHelper
```
Comment (meiersi-da) [inline ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml:34]: Consider renaming to `DivulgenceProposal` and replacing `Sign` with `Accept`.

**Review by meiersi-da — APPROVED** [review]
LGTM. Minor comments only.

File: ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml (hunk 7 above)
Comment (tudor-da) [reply ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml:30]: Done

File: ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml (hunk 6 above)
Comment (tudor-da) [reply ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml:27]: Done

File: ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml (hunk 8 above)
Comment (tudor-da) [reply ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml:19]: Done

File: ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml (hunk 5 above)
Comment (tudor-da) [reply ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml:64]: Done

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala (hunk 1 above)
Comment (tudor-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala]: Done

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala (hunk 2 above)
Comment (tudor-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala]: Done

File: ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml (hunk 3 above)
Comment (tudor-da) [reply ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml]: Done

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala (hunk 4 above)
Comment (tudor-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala]: Done

File: ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml (hunk 9 above)
Comment (tudor-da) [reply ledger/test-common/src/main/daml/semantic/DivulgenceTests.daml]: Done

**Pull request merged**